core/common/src/main/java/alluxio/underfs/ObjectLowLevelOutputStream.java

```diff
@@ -0,0 +1,402 @@
+    mPartitionSize = Math.max(UPLOAD_THRESHOLD, streamingUploadPartitionSize);
+    mPartNumber = new AtomicInteger(1);
+    if (ufsConf.isSet(PropertyKey.UNDERFS_OBJECT_STORE_STREAMING_UPLOAD_PART_TIMEOUT)) {
+      mUploadPartTimeout =
+          ufsConf.getDuration(PropertyKey.UNDERFS_OBJECT_STORE_STREAMING_UPLOAD_PART_TIMEOUT);
+    }
+  }
+
+  @Override
+  public void write(int b) throws IOException {
+    mSingleCharWrite[0] = (byte) b;
+    write(mSingleCharWrite);
+  }
+
+  @Override
+  public void write(byte[] b) throws IOException {
+    write(b, 0, b.length);
+  }
+
+  @Override
+  public void write(byte[] b, int off, int len) throws IOException {
+    if (b == null || len == 0) {
+      return;
+    }
+    validateWriteArgs(b, off, len);
+    if (mFile == null) {
+      initNewFile();
+    }
+    if (mPartitionOffset + len <= mPartitionSize) {
+      mLocalOutputStream.write(b, off, len);
+      mPartitionOffset += len;
+    } else {
+      int firstLen = (int) (mPartitionSize - mPartitionOffset);
+      mLocalOutputStream.write(b, off, firstLen);
+      mPartitionOffset += firstLen;
+      uploadPart();
+      write(b, off + firstLen, len - firstLen);
+    }
+  }
+
+  @Override
+  public void flush() throws IOException {
+    if (!isMultiPartUploadInitialized()) {
+      return;
+    }
+    // We try to minimize the time use to close()
+    // because Fuse release() method which calls close() is async.
+    // In flush(), we upload the current writing file if it is bigger than 5 MB,
+    // and wait for all current upload to complete.
+    if (mLocalOutputStream != null) {
+      mLocalOutputStream.flush();
+    }
+    if (mPartitionOffset > UPLOAD_THRESHOLD) {
+      uploadPart();
+    }
+    waitForAllPartsUpload();
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (mClosed) {
+      return;
+    }
+
+    // Set the closed flag, we never retry close() even if exception occurs
+    mClosed = true;
+
+    // Multi-part upload has not been initialized
+    if (!isMultiPartUploadInitialized()) {
+      if (mFile == null) {
+        LOG.debug("Streaming upload output stream closed without uploading any data.");
+        RetryUtils.retry("put empty object for key" + mKey, () -> createEmptyObject(mKey),
+            mRetryPolicy);
+      } else {
+        try {
+          mLocalOutputStream.close();
+          final String md5;
+          if (mHash != null) {
+            md5 = Base64.encodeBase64String(mHash.digest());
+          } else {
+            md5 = null;
+          }
+          RetryUtils.retry("put object for key" + mKey, () -> putObject(mKey, mFile, md5),
+              mRetryPolicy);
+        } finally {
+          if (!mFile.delete()) {
+            LOG.error("Failed to delete temporary file @ {}", mFile.getPath());
+          }
+        }
+      }
+      return;
+    }
+
+    try {
+      if (mFile != null) {
+        mLocalOutputStream.close();
+        int partNumber = mPartNumber.getAndIncrement();
```

Comment:
uploadPart already mPartNumber.getAndIncrement()?
can we add some test cases cover the part number changes?

Reply:
Added tests related to part number.